### PR TITLE
conditional compilation of ISA targets

### PR DESCRIPTION
**Pull request by dinfuehr**
Cretonne does not need to compile all available ISAs. This is useful for
e.g. JIT compilers, which only need to support the native ISA.

Targets can be specified with the environment variable `CRETONNE_TARGETS`.
There are three supported values:

1. `all` - compile all ISAs
2. `native` - compile only current ISA
3.  non-empty comma-separated list of ISAs

implements #13

### CODE REVIEW

**Comment by stoklund** [comment]
Thanks for working on this! I'll review this later today. 

Would it make sense to control this with cargo features instead of an environment variable? I imagine somebody using the Cretonne crate would just want to request a "native" feature in their Cargo.toml.

**Comment by dinfuehr** [comment]
Sure, sounds definitely the better solution. I will look into it.

File: lib/cretonne/build.rs [hunk 1]
```diff
@@ -44,3 +46,55 @@ fn main() {
...
+            }
+        }
+
+        "native" => {
+            let target = env::var("TARGET").expect("The TARGET environment variable must be set");
+            let arch = target.split("-").next().unwrap();
+
+            let isa = match arch {
+                "aarch64" => "arm64",
+                "arm" | "armv7" | "armv7s" => "arm32",
```
Comment (stoklund) [inline lib/cretonne/build.rs]: Maybe we should follow [Clang's triple parser](http://www.llvm.org/docs/doxygen/html/Triple_8cpp_source.html#l00335) here? There's a number of ARM variants, so it just defers to `starts_with("arm")` (or thumb).

File: lib/cretonne/build.rs [hunk 2]
```diff
@@ -44,3 +46,55 @@ fn main() {
...
+                "x86_64" | "i386" | "i586" | "i686" => "intel",
+                "riscv" => "riscv",
+                _ => {
+                    println!("no supported isa found for target triple `{}`", target);
+                    process::exit(1);
```
Comment (stoklund) [inline lib/cretonne/build.rs]: Did you check that this message actually appears on screen? Cargo normally swallows the build script output. Should it be written to stderr instead?

File: lib/cretonne/build.rs [hunk 3]
```diff
@@ -44,3 +46,55 @@ fn main() {
...
+            let mut found = false;
+
+            for isa in config.split(',') {
+                if possible_isas.contains(&isa) {
+                    println!("cargo:rustc-cfg=build_{}", isa);
+                    found = true;
+
+                } else {
+                    println!("unknown isa `{}`", isa);
+                    process::exit(1);
```
Comment (stoklund) [inline lib/cretonne/build.rs]: Maybe we should have a `fn fail(msg: String) -> !` function?

File: lib/cretonne/src/isa/mod.rs [hunk 4]
```diff
@@ -45,41 +45,83 @@ pub use isa::registers::{RegUnit, RegBank, RegInfo};
 use settings;
 use ir::{InstructionData, DataFlowGraph};
 
+#[cfg(build_riscv)]
 pub mod riscv;
```
Comment (stoklund) [inline lib/cretonne/src/isa/mod.rs]: Nice. This is exactly what I had in mind.

File: lib/cretonne/src/isa/mod.rs [hunk 5]
```diff
@@ -45,41 +45,83 @@ pub use isa::registers::{RegUnit, RegBank, RegInfo};
...
-    Some(riscv::isa_builder())
+#[cfg(build_riscv)]
+fn riscv_builder() -> Result<Builder, LookupError> {
+    Ok(riscv::isa_builder())
+}
+
+#[cfg(not(build_riscv))]
+fn riscv_builder() -> Result<Builder, LookupError>{
+    Err(LookupError::Unsupported)
+}
```
Comment (stoklund) [inline lib/cretonne/src/isa/mod.rs]: Yep. This was the reason I split out these little per-target functions. If you can think of a better way of implementing this, I'd be very interested.

Comment (dinfuehr) [inline lib/cretonne/src/isa/mod.rs]: Hmm, if you want we could create all those functions at compile-time with Python.

**Comment by dinfuehr** [comment]
Thanks for your feedback! I will try to address your points over the next few days.

File: lib/cretonne/src/isa/mod.rs (hunk 5 above)
Comment (stoklund) [inline lib/cretonne/src/isa/mod.rs]: That's probably not worth it. It's not that often we add a new target ISA, so it's OK if you have to write some boilerplate.

Let's just keep your functions. It's not too bad.

File: lib/reader/src/error.rs [hunk 6]
```diff
@@ -19,6 +19,8 @@ pub struct Error {
     pub location: Location,
     /// Error message.
     pub message: String,
+    /// Error because of unsupported ISA
+    pub unsupported: bool,
```
Comment (stoklund) [inline lib/reader/src/error.rs]: Hmm. I think it shouldn't be an error if a test file contains an unsupported ISA.

A test file [can contain multiple ISA specs](http://cretonne.readthedocs.io/en/latest/testing.html#file-tests) and it is possible that some of them are supported while others are unsupported. In that case, we should run the test on the supported ISAs and ignore the unsupported ISAs.

I think the `IsaSpec` enum can represent everything we need:

- The file has no ISA specs: Use `IsaSpec::None(..)`.
- The file has ISA specs, but none are supported: Use `IsaSpec::Some([])`.
- The file has ISA specs, not all supported: Use `IsaSpec::Some([a, b, ...])` with just the supported ISAs. Any unsupported ISA specs are ignored and omitted.

It should still be a parser error if we see an *unknown* ISA.

**Comment by stoklund** [comment]
I'm going to close this since @zummenix has taken over in #126, based in this work.

Thanks for getting us this far, @dinfuehr !

**Pull request closed without merging**